tests/test_operation.py

```diff
@@ -14,206 +14,183 @@
-            op_test(getattr(qml.ops, cls))
+        ww = list(range(cls.num_wires))
 
-        for cls in qml.ops._qubit__obs__:
-            op_test(getattr(qml.ops, cls))
+        # fixed parameter values
+        if cls.par_domain == 'A':
+            if cls.__name__ == "Interferometer":
+                ww = list(range(2))
+                par = [np.array([[0.83645892-0.40533293j, -0.20215326+0.30850569j],
+                                 [-0.23889780-0.28101519j, -0.88031770-0.29832709j]])]
+            else:
+                par = [np.array([[-1.82624687]])] * cls.num_params
+        else:
+            par = [-0.069125, 0.51778, 0.91133, 0.95904][:cls.num_params]
+
+        op = cls(*par, wires=ww, do_queue=False)
+
+        if issubclass(cls, qml.operation.Observable):
+            Q = op.heisenberg_obs(0)
+            # ev_order equals the number of dimensions of the H-rep array
+            assert Q.ndim == cls.ev_order
+            return
+
+        # not an Expectation
+
+        U = op.heisenberg_tr(num_wires=2)
+        I = np.eye(*U.shape)
+        # first row is always (1,0,0...)
+        assert np.all(U[0, :] == I[:, 0])
+
+        # check the inverse transform
+        V = op.heisenberg_tr(num_wires=2, inverse=True)
+        assert np.linalg.norm(U @ V -I) == pytest.approx(0, abs=tol)
+        assert np.linalg.norm(V @ U -I) == pytest.approx(0, abs=tol)
+
+        if op.grad_recipe is not None:
+            # compare gradient recipe to numerical gradient
+            h = 1e-7
+            U = op.heisenberg_tr(0)
+            for k in range(cls.num_params):
+                D = op.heisenberg_pd(k)  # using the recipe
+                # using finite difference
+                op.params[k] += h
+                Up = op.heisenberg_tr(0)
+                op.params = par
+                G = (Up-U) / h
+                assert D == pytest.approx(G, abs=tol)
+
+        # make sure that `heisenberg_expand` method receives enough wires to actually expand
+        # when supplied `wires` value is zero, returns unexpanded matrix instead of raising Error
+        # so only check multimode ops
+        if len(op.wires) > 1:
+            with pytest.raises(ValueError, match='is too small to fit Heisenberg matrix'):
+                op.heisenberg_expand(U, len(op.wires) - 1)
+
+        # validate size of input for `heisenberg_expand` method
+        with pytest.raises(ValueError, match='Heisenberg matrix is the wrong size'):
+            U_wrong_size = U[1:, 1:]
+            op.heisenberg_expand(U_wrong_size, len(op.wires))
+
+        # ensure that `heisenberg_expand` raises exception if it receives an array with order > 2
+        with pytest.raises(ValueError, match='Only order-1 and order-2 arrays supported'):
+            U_high_order = np.array([U] * 3)
+            op.heisenberg_expand(U_high_order, len(op.wires))
+
+
+
+    @pytest.mark.parametrize("cls", op_classes)
+    def test_operation_init(self, cls):
```

Comment:
What's wrong with this test?

Reply:
It loops over all operations, attempting to 'fix up' differences on a case-by-case basis with complicated branching. Since there are assertions nested within the conditional branching, the behavior of the test changes with the parametrized data, which makes it really hard to maintain/work with.

Ideally, there should be individual tests for each branch; e.g.,

* too many parameters
* too few parameters
* too many or too few wires
* repeated wires, etc.

Reply:
You'd still need the branching, that would just split the test into several. I don't think it helps much, since pytest will give you the exact assert that failed (and eventually they all must succeed).